File: 50.pow-x-n.py
#
# @lc app=leetcode.cn id=50 lang=python3
#
# [50] Pow(x, n)
#

# @lc code=start


class Solution:
    def myPow(self, x: float, n: int) -> float:
        if n < 0:
            x = 1 / x
            n = -n
        result = 1

        for i in range(n):
            result = result * x

        return result


s = Solution()
r = s.myPow(-2, 2)
print('r ->', r)


# https: // leetcode.cn/problems/powx-n/solutions/
class Solution:
    def myPow(self, x: float, n: int) -> float:
        def quickMul(N):
            if N == 0:
                return 1.0
            y = quickMul(N // 2)
            print('N // 2 ->', N // 2)
            return y * y if N % 2 == 0 else y * y * x

        return quickMul(n) if n >= 0 else 1.0 / quickMul(-n)


s = Solution()
r = s.myPow(2,10)
print('r ->', r)


class Solution:
    def myPow(self, x, n):
        if n < 0:
            x = 1 / x
            n = -n
        result = 1
        while n:
            # n & 1 的结果为0, 说明n是偶数
            # n & 1 的结果为1, 说明n是奇数
            if n & 1:
                result *= x
            x *= x
            n >>= 1
        return result


s = Solution()
r = s.myPow(0.00001, 2147483647)
print('r ->', r)
# @lc code=end

File: 50.pow-x-n.js
/*
 * @lc app=leetcode.cn id=50 lang=javascript
 *
 * [50] Pow(x, n)
 */

// @lc code=start
/**
 * @param {number} x
 * @param {number} n
 * @return {number}
 */
var myPow = function (x, n) {
  if (n < 0) {
    x = 1 / x
    n = -n
  }
  let result = 1

  for (let i = 0; i < n; i++) {
    result = result * x
  }
  return result
};

var myPow = function (x, n) {

  let helper = function (N) {
    if (N == 0) {
      return 1
    }
    let y = helper(Math.floor(N / 2))

    return N % 2 == 0 ? y * y : y * y * x
  }
  return n < 0 ? 1 / helper(-n) : helper(n)
};


var myPow = function (x, n) {
  function helper(N) {
    if (N == 0) {
      return 1
    }
    let y = helper(N)
    return N % 2 == 0 ? y * y : y * y * x
  }
  return n < 0 ? 1 / helper(-n) : helper(n)
};


var myPow = function (x, n) {
  if (n < 0) {
    n = -n
    x = 1 / x
  }
  let result = 1
  while (n) {
    if (n % 2 != 0) {
      result *= x
    }
    x *= x
    n = Math.floor(n / 2)
  }
  return result
};
myPow(2, 2)

// @lc code=end

